### ISSUE TITLE
Platform package install fails with Composer 2.2 lock file

### ISSUE THREAD

**Issue by ngrie** [heroku/heroku-buildpack-php#526]
Several people, including me, are experiencing an issue similar to #481 using the latest version of Composer. As far as I understand, it would be enough to bump the Composer version of this buildpack. I was able to work around the issue by setting the `plugin-api-version` to `2.1.0`.

```
!     > You are using Composer 1 which is deprecated. You should upgrade to Composer 2, see https://blog.packagist.com/deprecating-composer-1-support/
!     > Loading repositories with available runtimes and extensions
!     > Updating dependencies
!     > Your requirements could not be resolved to an installable set of packages.
!     >
!     >   Problem 1
!     >     - The requested package composer-plugin-api could not be found in any version, there may be a typo in the package name.
```

I am quite new to this, so did I understand correctly that future updates of Composer (which are automatically installed on my machine using homebrew) would break my deployments again as it happens today? If so, could/should that be fixed by improving the Composer version detection logic (I do not understand completely how that works yet)?

**Comment by beporter** [comment]
I just opened a Heroku support ticket for this same thing.

**Comment by boboldehampsink** [comment]
Same here

**Comment by beporter** [comment]
This section of the docs is probably relevant:

- https://github.com/heroku/heroku-buildpack-php/blob/653a6c795/support/build/README.md#how-heroku-installs-platform-dependencies

I understand that the early composer v1 run's goal is to "simulate" the later, real `composer install` early on to get extensions and PHP versions matching beforehand and ensure that second run succeeds, but wouldn't it make sense to make that first run a little more dynamic? Maybe detect the composer version to use for the first platform install run even?

I'm 100% sure there are dozens of other edge cases unknown to me currently, but maybe make this part of `bin/compile` smarter?

- https://github.com/heroku/heroku-buildpack-php/blob/653a6c795/bin/compile#L249,L251

**Comment by dzuelke** [comment]
FYI @beporter your analysis isn't quite correct - the problem was not that the platform install step runs using Composer 1, but that the Composer version stored in a new lock file is used as the version selector for the installation of the "real" Composer used later.

As y'all's lock files were generated using Composer 2.2.x, but that version wasn't yet available on Heroku, the dependency resolution failed.

It would have failed just the same way if that installation step had happened using Composer 2.1.x.

The use of Composer 1 for this step is just for bootstrapping - that step installs Composer 2 (or 1, if your lock file says so), PHP, extensions, all that stuff. What exact version of Composer is used in that step is not impacted by what version of Composer the user dependencies lock file was generated with.

(Still, we're currently moving that platform install step to Composer 2, which will solve a few long-standing edge cases such as https://github.com/phpseclib/bcmath_compat/issues/1, but Composer 2 will then also be used for Composer 1 user projects during platform installation - which will then pull in Composer 1 for the actual user dependencies `composer install`).

Anyway, this is fixed in #527, which also implemented forward compatibility for installations, so that when Composer 2.3.0 is released, lock files generated using that version will not fail to build in the same way until 2.3.0 is also available on Heroku.

**Comment by dzuelke** [comment]
Keep in mind that the initial platform install step isn't just a vanilla `composer install` on the generated platform `composer.json` - there is a custom Composer Installer plugin that does all the heavy lifting under the hood, and that plugin is currently written for Composer 1. That's why the platform install step always runs using Composer 1 right now (until the migration is complete), because it wouldn't work with Composer 2.

**Comment by jonnyeom** [comment]
I'm not sure why this is closed or if I am doing something wrong.
I am still running into this issue.


I have composer 2.2.3 running and composer.lock has the plugin api to 2.2.0
But the heroku build still defaults to composer 1.


Is there something else I have to do besides run `composer update`?
Do I have to wait for another release of the buildpack?

**Comment by jonnyeom** [comment]
The fix from #529 resolved the issue I was  running into.  
Big thanks!
